kube-client/src/api/portforward.rs

```diff
@@ -0,0 +1,225 @@
+use std::{
+    future::Future,
+    pin::Pin,
+    sync::{Arc, Mutex},
+    task::{Context, Poll, Waker},
+};
+
+use futures::{
+    channel::{mpsc, oneshot},
+    FutureExt, SinkExt, StreamExt,
+};
+use tokio::io::{AsyncRead, AsyncWrite, AsyncWriteExt, DuplexStream};
+use tokio_tungstenite::{tungstenite as ws, WebSocketStream};
+
+type ErrorReceiver = oneshot::Receiver<String>;
+type ErrorSender = oneshot::Sender<String>;
+
+enum Message {
+    FromPod(Vec<u8>),
+    ToPod(Vec<u8>),
+}
+
+struct PortforwarderState {
+    waker: Option<Waker>,
+    finished: bool,
+    ports: Option<Vec<Port>>,
```

Comment:
does this need to be option wrapped? users have to pass in a slice of ports, so the worst that happens here is that `ports.is_empty()`

Reply:
`ports` is never empty because it's required to be non empty in `Request::portforward`.

The reason why `ports` is an `Option` is so we can move it out (see `PortforwarderState::ports()`).

Reply:
it leads to guaranteed unwraps on the user side though (example line 39 and 40). don't you think we should at least do the unwrap for them if we know these always exists?

Reply:
> The reason why ports is an Option is so we can move it out (see Portforwarder::ports()).

i thought maybe we could return `&mut [Port]` via `self.ports.as_mut_slice()` and rely on rust ownership to ensure that there's only one user of each port.

Reply:
Yeah, `&mut [Port]` sounds good. I actually don't remember why I wanted to allow moving `ports` out of `Portforwarder`. It shouldn't be necessary.